plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/AllFilesTable.java

```diff
@@ -0,0 +1,172 @@
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+
+public class AllFilesTable implements SystemTable {
+    private final ConnectorTableMetadata tableMetadata;
+    private final Table icebergTable;
+
+    public AllFilesTable(SchemaTableName tableName, TypeManager typeManager, Table icebergTable)
+    {
+        this.icebergTable = requireNonNull(icebergTable, "icebergTable is null");
+        this.tableMetadata = new ConnectorTableMetadata(requireNonNull(tableName, "tableName is null"),
+                ImmutableList.<ColumnMetadata>builder()
+                        .add(new ColumnMetadata("content", INTEGER))
+                        .add(new ColumnMetadata("file_path", VARCHAR))
+                        .add(new ColumnMetadata("file_format", VARCHAR))
+                        .add(new ColumnMetadata("record_count", BIGINT))
+                        .add(new ColumnMetadata("file_size_in_bytes", BIGINT))
+                        .add(new ColumnMetadata("column_sizes", typeManager.getType(mapType(INTEGER.getTypeSignature(), BIGINT.getTypeSignature()))))
+                        .add(new ColumnMetadata("value_counts", typeManager.getType(mapType(INTEGER.getTypeSignature(), BIGINT.getTypeSignature()))))
+                        .add(new ColumnMetadata("null_value_counts", typeManager.getType(mapType(INTEGER.getTypeSignature(), BIGINT.getTypeSignature()))))
+                        .add(new ColumnMetadata("nan_value_counts", typeManager.getType(mapType(INTEGER.getTypeSignature(), BIGINT.getTypeSignature()))))
+                        .add(new ColumnMetadata("lower_bounds", typeManager.getType(mapType(INTEGER.getTypeSignature(), VARCHAR.getTypeSignature()))))
+                        .add(new ColumnMetadata("upper_bounds", typeManager.getType(mapType(INTEGER.getTypeSignature(), VARCHAR.getTypeSignature()))))
+                        .add(new ColumnMetadata("key_metadata", VARBINARY))
+                        .add(new ColumnMetadata("split_offsets", new ArrayType(BIGINT)))
+                        .add(new ColumnMetadata("equality_ids", new ArrayType(INTEGER)))
+                        .build());
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        return new FixedPageSource(buildPages());
+    }
+
+    @Override
+    public Distribution getDistribution() {
+        return Distribution.SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata() {
+        return tableMetadata;
+    }
+
+    protected List<Page> buildPages(long snapshotId)
+    {
+        PageListBuilder pagesBuilder = PageListBuilder.forTable(tableMetadata);
+        Map<Integer, Type> idToTypeMapping = getIcebergIdToTypeMapping(icebergTable.schema());
+
+        TableScan tableScan = icebergTable.newScan()
```

Comment:
Did you follow https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/AllDataFilesTable.java as reference implementation when implementing this functionality? I am seeing that the implementation doesn't go through all the manifests

Reply:
@findinpath My understanding is (please correct me if I'm wrong)

The current `$files` metadata table gets data files for a specific snapshot, by directly using `scan.planFiles` instead of first list manifests using `snapshot.allManifests()` and then read the DataFile entries from the manifest file.
If we see the current implementation of the `$file` is [here](https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java#L111), which is for a single snapshot.
And in the `$all_files`, we are using the same logic but iterating all the snapshots and getting file details.

Reply:
@osscm, you are correct.
When using a `DataTableScan` for doing the `planFiles` operation we'll be going through the manifests of the snapshot.
Your implementation does go through all the snapshots of the table, so I am assuming that all the files are actually collected.

Reply:
yes, @findinpath that's correct!